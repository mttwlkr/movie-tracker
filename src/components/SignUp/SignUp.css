button.submit {
  background-color: #8C8C8C;
  border: none;
  outline: none;
  border-radius: 20px;
  margin-top: 16px;
  padding: 10px;
  font-size: 16px;
}

button.submit:hover {
  background-color: #545454;
}