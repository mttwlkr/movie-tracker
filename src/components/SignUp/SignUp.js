import React, {Component} from 'react';
import { Redirect, NavLink } from 'react-router-dom';
import './SignUp.css';

class SignUp extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: '',
      password: '',
      authenticated: false,
      signUpError: false
    };
  }

  handleFetch = async (state) => {
    try {
      const response = await fetch('/api/users/new', {
        method: "POST",
        body: JSON.stringify(state),
        headers: {
          'content-type': 'application/json'
        }
      });
      const newAccount = await response.json();
     
      if (newAccount.error) {
        this.setState({ signUpError: true });
      } else {
        this.setState({ authenticated: true });
      }

    } catch (error) {
      alert(error);
    }
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({[name]: value});
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const info = {
      name: this.state.name,
      email: this.state.email,
      password: this.state.password
    };
    this.handleFetch(info);
    this.setState({
      name: '',
      email: '',
      password: ''
    });
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <input 
            type='text'
            name='name'
            id='name'
            value={this.state.name}
            placeholder='Name'
            onChange={this.handleChange}
          />
          <input 
            type='email'
            name='email'
            id='email'
            value={this.state.email}
            placeholder='Email'
            onChange={this.handleChange}
          />
          <input 
            type='password'
            name='password'
            id='password'
            value={this.state.password}
            placeholder='Password'
            onChange={this.handleChange}
          />
          <button className='submit'>Submit</button>
        </form>
        { 
          this.state.signUpError &&
            <section>
              <p>This email has already been used</p>
              <NavLink to="/login">Already a member?</NavLink>
            </section>
        }
        {
          this.state.authenticated && 
            <Redirect to='/login' />
        }
      </div>
    );
  }
}

export default SignUp;