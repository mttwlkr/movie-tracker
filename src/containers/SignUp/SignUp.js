import React, {Component} from 'react';
import { NavLink, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import './SignUp.css';
import { logInUser } from '../../actions';
import PropTypes from 'prop-types';

const url = process.env.REACT_APP_DATABASE_URL;

export class SignUp extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: '',
      password: '',
      signUpError: false
    };
  }
 
  handleFetch = async (userInfo) => { 
    try {
      const response = await fetch(`${url}/api/v1/users`, {
        method: "POST",
        body: JSON.stringify(userInfo),
        headers: {
          'content-type': 'application/json'
        }
      });

      if (!response.ok) {
        throw new Error(response.statusText);
      }

      const newAccount = await response.json();
      const { id, name } = newAccount;

      this.props.logInUser(id, name);
      this.props.history.push('/');
      
    } catch (error) {
      this.setState({ signUpError: true });
    }
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({[name]: value});
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const info = {
      name: this.state.name,
      email: this.state.email,
      password: this.state.password
    };
    this.handleFetch(info);
    this.setState({
      name: '',
      email: '',
      password: ''
    });
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <input 
            type='text'
            name='name'
            id='name'
            value={this.state.name}
            placeholder='Name'
            onChange={this.handleChange}
          />
          <input 
            type='email'
            name='email'
            id='email'
            value={this.state.email}
            placeholder='Email'
            onChange={this.handleChange}
          />
          <input 
            type='password'
            name='password'
            id='password'
            value={this.state.password}
            placeholder='Password'
            onChange={this.handleChange}
          />
          <button className='submit'>Submit</button>
        </form>
        { 
          this.state.signUpError &&
            <section>
              <p>This email has already been used</p>
              <NavLink to="/login">Already a member?</NavLink>
            </section>
        }
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    logInUser: (id, name) => (dispatch(logInUser(id, name)))
  };
};

export default withRouter(connect(null, mapDispatchToProps)(SignUp));

SignUp.propTypes = {
  logInUser: PropTypes.func,
  history: PropTypes.object
};