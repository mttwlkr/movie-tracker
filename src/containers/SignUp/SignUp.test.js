import React from 'react';
import { SignUp, mapDispatchToProps } from './SignUp';
import { shallow } from 'enzyme';

describe('SignUp', () => {
  let wrapper;
  const mockLogInUser = jest.fn();
  const mockHistory = { push: jest.fn() };
  const mockUserData = {
    name: 'Matt',
    password: 'yolo',
    email: 'test@test.com'
  };

  beforeEach(() => {
    wrapper = shallow(<SignUp logInUser={mockLogInUser} history={mockHistory} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleFetch', () => {
    it('should call logInUser if response is ok and redirect', async () => {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: true,
        json: () => Promise.resolve({
          name: 'Matt',
          id: 3
        })
      }));
  
      await wrapper.instance().handleFetch(mockUserData);
      expect(mockLogInUser).toHaveBeenCalledWith(3, 'Matt');
      expect(mockHistory.push).toHaveBeenCalledWith('/');
    });
  
    it('should signUpError state to true if response is bad', async () => {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: false,
        statusText: 'Not Found'
      }));
  
      expect(wrapper.state('signUpError')).toEqual(false);
      await wrapper.instance().handleFetch(mockUserData);
      expect(wrapper.state('signUpError')).toEqual(true);
    });
  });

  describe('handleChange', () => {
    it('should update name, email and password when input changes', () => {
      const mailEvent = { 
        target: { 
          value: 'terminator@dude.com', 
          name: 'email' 
        } 
      };
  
      const nameEvent = { 
        target: { 
          value: 'Arnie', 
          name: 'name' 
        } 
      };
  
      const passwordEvent = { 
        target: { 
          value: 'doomsday', 
          name: 'password' 
        } 
      };
     
      wrapper.find('#name').simulate('change', nameEvent);
      expect(wrapper.state('name')).toEqual('Arnie');
  
      wrapper.find('#email').simulate('change', mailEvent);
      expect(wrapper.state('email')).toEqual('terminator@dude.com');
  
      wrapper.find('#password').simulate('change', passwordEvent);
      expect(wrapper.state('password')).toEqual('doomsday');
    });  
  });

  describe('handleSubmit', () => {
    it('should invoke handleFetch when handleSubmit is called', () => {
      const mockEvent = { preventDefault: jest.fn() };
      const mockInfo = {
        name: 'Bob',
        email: 'bobbafetch@dude.com',
        password: 'boobbafette'
      };
  
      wrapper.instance().handleFetch = jest.fn();
      const spy = jest.spyOn(wrapper.instance(), 'handleFetch');
  
      wrapper.setState({
        name: 'Bob',
        email: 'bobbafetch@dude.com',
        password: 'boobbafette'
      });
  
      wrapper.instance().handleSubmit(mockEvent);
      expect(spy).toHaveBeenCalledWith(mockInfo);
    });
  
    it('should reset state to empty when handleSubmit is called', async () => {
  
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.setState({
        name: 'Matt', 
        password: 'sploosh', 
        email: 'ohmama@zguy.com'
      });
  
      const expected = {
        name: '',
        email: '',
        password: '',
        signUpError: false
      };
      wrapper.instance().handleFetch = jest.fn();
  
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.state()).toEqual(expected);
    });
  });

  it('should dispatch to the store correctly', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.logInUser();
    expect(mockDispatch).toHaveBeenCalled();
  });
});