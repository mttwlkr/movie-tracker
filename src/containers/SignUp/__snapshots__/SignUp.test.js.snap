// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUp
    history={
        Object {
            "push": [Function],
          }
    }
    logInUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
>
          <input
                    id="name"
                    name="name"
                    onChange={[Function]}
                    placeholder="Name"
                    type="text"
                    value=""
          />
          <input
                    id="email"
                    name="email"
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
          />
          <input
                    id="password"
                    name="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
          />
          <button
                    className="submit"
          >
                    Submit
          </button>
</form>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              id="name"
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
/>,
            <input
              id="email"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              type="email"
              value=""
/>,
            <input
              id="password"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
/>,
            <button
              className="submit"
>
              Submit
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "name",
              "name": "name",
              "onChange": [Function],
              "placeholder": "Name",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "email",
              "name": "email",
              "onChange": [Function],
              "placeholder": "Email",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "password",
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
>
            <input
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Name"
                        type="text"
                        value=""
            />
            <input
                        id="email"
                        name="email"
                        onChange={[Function]}
                        placeholder="Email"
                        type="email"
                        value=""
            />
            <input
                        id="password"
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
            />
            <button
                        className="submit"
            >
                        Submit
            </button>
</form>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
/>,
              <input
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="email"
                value=""
/>,
              <input
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
/>,
              <button
                className="submit"
>
                Submit
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "name",
                "name": "name",
                "onChange": [Function],
                "placeholder": "Name",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "email",
                "name": "email",
                "onChange": [Function],
                "placeholder": "Email",
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "password",
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "className": "submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
